### ISSUE TITLE
`int_range()` gives incorrect results with expressions that evaluate to negative numbers

### ISSUE THREAD

**Issue by Wainberg** [pola-rs/polars#10849]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

In each example, the first one is correct, but the second one gives an overflow error because `len()` and `null_count()` return unsigned integers. Each pair should give the same result.

```python
>>> pl.DataFrame({'a': [1]}).select(pl.int_range(0, -1, -1))
 int
 0
shape: (0, 1)
>>> pl.DataFrame({'a': [1]}).select(pl.int_range(0, -pl.col('a').len(), -1))
thread '<unnamed>' panicked at crates/polars-core/src/chunked_array/arithmetic/numeric.rs:229:54:
attempt to subtract with overflow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "polars/py-polars/polars/dataframe/frame.py", line 8097, in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "polars/py-polars/polars/lazyframe/frame.py", line 1730, in collect
    return wrap_df(ldf.collect())
                   ^^^^^^^^^^^^^
pyo3_runtime.PanicException: attempt to subtract with overflow
```

```python
>>> pl.DataFrame({'a': [None]}).select(pl.int_range(0, -1, -1))
 int
 0
shape: (0, 1)
>>> pl.DataFrame({'a': [None]}).select(pl.int_range(0, -pl.col('a').null_count(), -1))
thread '<unnamed>' panicked at crates/polars-core/src/chunked_array/arithmetic/numeric.rs:229:54:
attempt to subtract with overflow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "polars/py-polars/polars/dataframe/frame.py", line 8097, in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "polars/py-polars/polars/lazyframe/frame.py", line 1730, in collect
    return wrap_df(ldf.collect())
                   ^^^^^^^^^^^^^
pyo3_runtime.PanicException: attempt to subtract with overflow
```

```python
>>> pl.DataFrame({'a': [1]}).select(pl.int_range(-1, 1))
 int
 -1
 0
shape: (2, 1)
>>> pl.DataFrame({'a': [1]}).select(pl.int_range(-pl.col('a').len(), 1))
thread '<unnamed>' panicked at crates/polars-core/src/chunked_array/arithmetic/numeric.rs:229:54:
attempt to subtract with overflow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "polars/py-polars/polars/dataframe/frame.py", line 8097, in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "polars/py-polars/polars/lazyframe/frame.py", line 1730, in collect
    return wrap_df(ldf.collect())
                   ^^^^^^^^^^^^^
pyo3_runtime.PanicException: attempt to subtract with overflow
```

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.2
Index type:           UInt32
Platform:             Linux-4.4.0-22621-Microsoft-x86_64-with-glibc2.35
Python:               3.12.1 | packaged by conda-forge | (main, Dec 23 2023, 08:03:24) [GCC 12.3.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
matplotlib:           3.8.2
numpy:                1.26.2
openpyxl:             3.1.2
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             0.8.1
xlsxwriter:           3.1.9
```
</details>

**Comment by orlp** [comment]
So the problem here is that `pl.col('a').len()` and `pl.col('a').null_count()` return unsigned integers, which when negated result in huge positive numbers:

```python
>>> pl.DataFrame({'a': [1, 2, 5]}).select(-pl.col('a').len())
shape: (1, 1)
┌────────────┐
│ literal    │
│ ---        │
│ u32        │
╞════════════╡
│ 4294967293 │
└────────────┘
```

Honestly we should probably either disallow `-` on unsigned integers, or have it implicitly cast to a signed integer first.

**Comment by Wainberg** [comment]
What about just using signed integers? That's what pl.Series seems to already use:

```python
>>> -pl.Series('a', [1]).len()
-1
>>> pl.DataFrame({'a': [1]}).select(-pl.col('a').len())
 literal
 4294967295
shape: (1, 1)
```

**Comment by Wainberg** [comment]
What's going on here, seems like a separate issue?

```python
>>> pl.DataFrame({'a': [1, 2, 5]}).select(pl.int_range(pl.col('a').is_not_null().sum(), -1, -1))
Traceback (most recent call last):
[...]
exceptions.NoDataError: no data in `end` evaluation
>>> pl.DataFrame({'a': [1, 2, 5]}).select(pl.int_range(pl.col('a').is_not_null().sum().cast(pl.UInt64), -1, -1))
 int
 3
 2
 1
 0
 -1
shape: (5, 1)
>>> pl.DataFrame({'a': [1, 2, 5]}).select(pl.int_range(pl.col('a').is_not_null().sum().cast(pl.Int64), -1, -1))
 int
 3
 2
 1
 0
 -1
shape: (5, 1)
```

**Comment by orlp** [comment]
@Wainberg 

> What about just using signed integers? That's what pl.Series seems to already use:

```
>>> -pl.Series('a', [1]).len()
-1
```

No, that's still an unsigned integer inside polars, that gets returned as a Python `int` (which has no concept of unsigned integers) and *then* it gets negated inside Python.

**Comment by Wainberg** [comment]
I suppose it's fine to store them as unsigned integers internally, and convert to signed whenever necessary - in other words, if the user can use them as if they're signed integers and always get the expected result.

Thoughts on https://github.com/pola-rs/polars/issues/10849#issuecomment-1702447741?

**Comment by orlp** [comment]
@Wainberg https://github.com/pola-rs/polars/issues/10849#issuecomment-1702447741 is rather strange, I made that a dedicated issue: https://github.com/pola-rs/polars/issues/10867

I also noticed that the range is incorrectly not making the exclusive bound... exclusive for negative steps. So I made another issue for that: https://github.com/pola-rs/polars/issues/10865.

Now neither of those issues is your issue, which is due to bad casting of negative unsigned numbers. Next week I'll speak a bit more with the team to see how we want to handle unsigned number negation.
